### PR TITLE
libjsc: convert to jansson and cleanup

### PR DESCRIPTION
**Pull request by garlick**
This PR converts libjsc from json-c to jansson by pulling in `shortjansson.h` from flux-sched.

It then begins to simplify the code a bit using `json_pack()`, `json_unpack()`, etc..

Along the way some back to back KVS RPCs are able to be parallelized.

I'm still working on this - just posting a checkpoint in case @dongahn wants to have an early peek and see if I'm going off track in any way.

### CODE REVIEW

**Comment by dongahn** [comment]
@garlick: I just gave a cursory look and found lots of neat changes! 

`json_pack ()` does improve readability a lot.

In terms of optimization, b23d23c seems to show the general direction of using non-blocking asynchronous communications (kvs in this case). This particular change exposes and exploits KVS parallelism within one function.

This makes me wonder if you also plan to expose parallelism across different handlers as part of JSC refactoring? I haven't thought much about this and not sure if JSC has enough logic to benefit from this way, but I was wondering if you are thinking about this.

**Comment by garlick** [comment]
Just rebased on current master and fixed a couple of uninitialized variables in error paths that travis caught.

@dongahn - in this PR I mainly wanted to convert to jansson, and simplify the code a bit by using some of the higher level jansson API calls that we didn't have in json-c.  The parallelism in the JSC_RDESC query, and the a similar opportunity in the JSC_PDESC were about as far as I thought was appropriate in this PR since it allows the API to remain the same.

In other words when you call `jsc_query_jcb()` it still blocks internally waiting for KVS RPC(s).

In a future PR (or in the update for new exec system) we should definitely look at having the query/update functions return futures IMHO.

**Comment by grondo** [comment]
Restarted a build that stalled after the issues tests.

**Comment by grondo** [comment]
This is great cleanup, LGTM. What else remains to be done before we can merge it?

**Comment by dongahn** [comment]
That sounds good. I was just wondering your future plan. Thanks.

**Comment by garlick** [comment]
Well there is one more big one, the JSC_PDESC query, but actually this could be merged as is to save me from getting distracted for another day :-)

It does need a test with flux-sched first though.

**Comment by garlick** [comment]
Sched 0.5.0 does pass `make check` when built against this PR's branch, once the unrelated patch in flux-framework/flux-core#345 was applied to address a 32-bit build system issue.

**Comment by dongahn** [comment]
> Sched 0.5.0 does pass make check when built against this PR's branch

Nice job!

**Comment by grondo** [comment]
I notice a couple static functions don't have sanity coverage in the flux-core tests. Probably can't do anything about the rdl query function, but wreckrun does set a ersatz `R_lite` when running without the scheduler, so this addition to `t2001-jsc.t` might be warranted:

```diff
diff --git a/t/t2001-jsc.t b/t/t2001-jsc.t
index ea2109c..e88ab87 100755
--- a/t/t2001-jsc.t
+++ b/t/t2001-jsc.t
@@ -199,6 +199,10 @@ test_expect_success 'jstat 7.5: basic query works: pdesc' '
     flux jstat query 1 pdesc
 '
 
+test_expect_success 'jstat 7.6: basic query works: R_lite' '
+    flux jstat query 1 R_lite
+'
+
 test_expect_success 'jstat 8: query detects bad inputs' '
     test_expect_code 42 flux jstat query 0 jobid &&
     test_expect_code 42 flux jstat query 99999 state-pair &&

```

**Comment by garlick** [comment]
Rebased and also dropped shortjson.h and libjson-c now that we're free of it.

That ought to reduce my lines-of-code output for the year!

**Comment by dongahn** [comment]
I don't know if we want to handle the python jsc binding as part of this PR: https://github.com/flux-framework/flux-core/blob/master/src/bindings/python/flux/jsc.py. But this binding is a bit out of date -- e.g., deprecated keys are still there etc. 

I am okay if we want to handle python in a separate PR. But at some point we need to revisit this as it looks to me like the initial MLSI users are python users.

**Comment by garlick** [comment]
Sounds like a separate PR to me, since the libjsc interfaces did not change in the PR (except for adding ngpus to the one function that was added for @trws.

IT would be good to find out what interfaces are required for this particular application and see if we can provide it in a way that can transition to the new exec system.  New issue?

**Comment by dongahn** [comment]
Agreed.

> IT would be good to find out what interfaces are required for this particular application and see if we can provide it in a way that can transition to the new exec system. New issue?

I should let this user explore the basic functionality first. Then we can move to advanced feature, I will create an issue or issues to track this. Compatibility with the new exec system is a decent goal!

**Comment by grondo** [comment]
Nice work!

I restarted the coverage builder because for some reason it is reporting that the build failed. However, Travis did report green in all builds. Is this ready to merge?

**Comment by trws** [comment]
This looks really good to me.  The python stuff we should be able to 
clean up in another ticket.  It may be possible to do substantially 
better by the bindings now that it supports importing some defines.  
Alternately, we could actually make it a global constant struct that 
contains all those values then the bindings would just pick it up, 
either way.

**Comment by garlick** [comment]
Thanks!

Yes merge please.
